packages/core/src/signals/src/computed.ts

```diff
@@ -0,0 +1,161 @@
+  const node = new ComputedImpl(computation, equal);
+  return markSignal(node.signal.bind(node));
+}
+
+/**
+ * A dedicated symbol used before a computed value has been calculated for the first time.
+ */
+const UNSET: any = Symbol('UNSET');
+
+/**
+ * A dedicated symbol used in place of a computed signal value to indicate that a given computation
+ * is in progress. Used to detect cycles in computation chains.
+ */
+const COMPUTING: any = Symbol('COMPUTING');
+
+/**
+ * A dedicated symbol used in place of a computed signal value to indicate that a given computation
+ * failed. The thrown error is cached until the computation gets dirty again.
+ */
+const ERRORED: any = Symbol('ERRORED');
+
+/**
+ * A computation, which derives a value from a declarative reactive expression.
+ *
+ * `Computed`s are both `Producer`s and `Consumer`s of reactivity.
+ */
+class ComputedImpl<T> implements Producer, Consumer {
+  readonly id = nextReactiveId();
+  readonly ref = new WeakRef(this);
+  readonly producers = new Map<ProducerId, Edge>();
+  readonly consumers = new Map<ConsumerId, Edge>();
+  trackingVersion = 0;
+  valueVersion = 0;
+
+  /**
+   * Current value of the computation.
+   *
+   * This can also be one of the special values `UNSET`, `COMPUTING`, or `ERRORED`.
+   */
+  value: T = UNSET;
+
+  /**
+   * If `value` is `ERRORED`, the error caught from the last computation attempt which will
+   * be re-thrown.
+   */
+  error: unknown = null;
+
+  /**
+   * Flag indicating that the computation is currently stale, meaning that one of the
+   * dependencies has notified of a potential change.
+   *
+   * It's possible that no dependency has _actually_ changed, in which case the `stale`
+   * state can be resolved without recomputing the value.
+   */
+  stale = true;
+
+  constructor(private computation: () => T, private equal: (oldValue: T, newValue: T) => boolean) {}
+
+  checkForChangedValue(): void {
+    if (!this.stale) {
+      // The current value and its version are already up to date.
+      return;
+    }
+
+    // The current value is stale. Check whether we need to produce a new one.
+
+    if (this.value !== UNSET && this.value !== COMPUTING && !consumerPollValueStatus(this)) {
+      // Even though we were previously notified of a potential dependency update, all of
+      // our dependencies report that they have not actually changed in value, so we can
+      // resolve the stale state without needing to recompute the current value.
+      this.stale = false;
+      return;
+    }
+
+    // The current value is stale, and needs to be recomputed. It still may not change -
+    // that depends on whether the newly computed value is equal to the old.
+    this.recomputeValue();
+  }
+
+  private recomputeValue(): void {
+    if (this.value === COMPUTING) {
+      // Our computation somehow led to a cyclic read of itself.
+      throw new Error('Detected cycle in computations.');
```

Comment:
```suggestion
      throw Error('Detected cycle in computations.');
```
optional: `throw Error(...)` is equivalent

Reply:
FWIW, the code base uses mostly `throw new Error('...')`

Reply:
I don't know exactly how, but maybe the error could be improved to mention the computation? It would be easier to debug when a cycle is introduced.

Reply:
The entire cycle should be in the stacktrace.